lib/iris/analysis/__init__.py

```diff
@@ -2338,180 +2347,117 @@ def group(self):
+                    # or whether we need to get the coordinates next group.
+                    if groupby_slice.stop == self._stop:
+                        # This coordinate has exhausted all its groups,
+                        # so remove it.
+                        groupby_slices[index] = None
+                    elif groupby_slice.stop == stop:
+                        # The current group of this coordinate is
+                        # exhausted, so get the next one.
+                        groupby_slices[index] = next(groups[index])
+
+            # Cache the indices
+            self._groupby_indices = [tuple(i) for i in indices_by_key.values()]
+            # Calculate the new group-by coordinates.
+            self._compute_groupby_coords()
+            # Calculate the new shared coordinates.
+            self._compute_shared_coords()
+
+        # Return the group-by indices/groups.
+        return self._groupby_indices
+
+    def _compute_groupby_coords(self) -> None:
         """Create new group-by coordinates given the group slices."""
-
-        groupby_slice = []
-
-        # Iterate over the ordered dictionary in order to construct
-        # a group-by slice that samples the first element from each group.
-        for key_slice in self._slices_by_key.values():
-            if isinstance(key_slice, tuple):
-                groupby_slice.append(key_slice[0])
-            else:
-                groupby_slice.append(key_slice.start)
-
-        groupby_slice = np.array(groupby_slice)
+        # Construct a group-by slice that samples the first element from each
+        # group.
+        groupby_slice = np.array([i[0] for i in self._groupby_indices])
 
         # Create new group-by coordinates from the group-by slice.
         self.coords = [coord[groupby_slice] for coord in self._groupby_coords]
 
-    def _compute_shared_coords(self):
+    def _compute_shared_coords(self) -> None:
         """Create the new shared coordinates given the group slices."""
-
-        groupby_indices = []
-        groupby_bounds = []
-
-        # Iterate over the ordered dictionary in order to construct a list of
-        # tuple group indices, and a list of the respective bounds of those
-        # indices.
-        for key_slice in self._slices_by_key.values():
-            if isinstance(key_slice, tuple):
-                indices = key_slice
-            else:
-                indices = tuple(range(*key_slice.indices(self._stop)))
-
-            groupby_indices.append(indices)
-            groupby_bounds.append((indices[0], indices[-1]))
-
-        # Create new shared bounded coordinates.
         for coord, dim in self._shared_coords:
             climatological_coord = (
                 self.climatological and coord.units.is_time_reference()
             )
             if coord.points.dtype.kind in "SU":
                 if coord.bounds is None:
-                    new_points = []
+                    new_points_list = []
```

Comment:
Unlike the empty lists in the constructor, in this case, [`mypy` can infer the type](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html) because the list is used inside this function. You can check by adding `reveal_type(new_points_list)` on line 2570. However, if you feel it improves readability, I would be happy to add type hints for it.

Reply:
No that's fine, I can see the distinction